addons/mrp/models/stock_picking.py

```diff
@@ -106,6 +115,29 @@ def action_detailed_operations(self):
         action['context']['has_kits'] = self.has_kits
         return action
 
+    @api.depends('group_id.stock_move_ids.created_production_id.procurement_group_id.mrp_production_ids',
+                'move_ids.move_dest_ids.raw_material_production_id')
+    def _compute_mrp_production_ids(self):
+        self.production_ids = self.env['mrp.production']._read_group(
+            domain=[('procurement_group_id', 'in', self.group_id.ids)],)
+        self.production_ids |= self.move_ids.move_dest_ids.raw_material_production_id
+        self.production_count = len(self.production_ids)
```

Comment:
I'm pretty sure the concern of @amoyaux was to avoid search in the for loop, not the loop itself :suspect: 
You have to keep the compute multi and you can't assign the result of a `_read_group` (a list) to a field like that.
No sure this suggestion converts all cases since it's based on procurement group only but it gives you the idea
```suggestion
    @api.depends('group_id')
    def _compute_mrp_production_ids(self):
        data = self.env['procurement.group'].read_group(
            domain=[('id', 'in', self.group_id.ids)],
            groupby=['id', 'mrp_production_ids'],
            fields=['mrp_production_ids'])
        groups_by_id = {group['id'].id: group['id'] for group in data}
        for picking in self:
            group = groups_by_id.get(picking.group_id.id, self.env['procurement.group'])
            picking.production_ids = group.mrp_production_ids
            picking.production_count = len(picking.production_ids)
```
If you need to find MOs thanks to their moves too, you can first search all MO you need, and then loop to assign them to each corresponding picking

Reply:
> I'm pretty sure the concern of @amoyaux was to avoid search in the for loop, not the loop itself :suspect: You have to keep the compute multi and you can't assign the result of a `_read_group` (a list) to a field like that. No sure this suggestion converts all cases since it's based on procurement group only but it gives you the idea
> 
> If you need to find MOs thanks to their moves too, you can first search all MO you need, and then loop to assign them to each corresponding picking

Hello @svs-odoo Thanks for your suggestion. But we have to add this line also: 
`group |= picking.move_ids.move_dest_ids.raw_material_production_id.procurement_group_id`
when we don't have a group in case of `procure_method` set to `take from stock`.
see ref video: https://app.screencastify.com/v2/watch/DtveZALgkxJiChD9s2wW
Let me know if I misunderstood.
Thanks!

Reply:
@jipr-odoo I'm not sure to understand why you have to take the group here if the goal is to find the production and you go through the production to get the group ?
Isn't simpler to stop add the `raw_material_production_id` ? 
```python
productions = picking.move_ids.move_dest_ids.raw_material_production_id
```

Reply:
> @jipr-odoo I'm not sure to understand why you have to take the group here if the goal is to find the production and you go through the production to get the group ? Isn't simpler to stop add the `raw_material_production_id` ?
> 
> ```python
> productions = picking.move_ids.move_dest_ids.raw_material_production_id
> ```

Hello @svs-odoo  I appreciate your suggestion, and I agree that focusing on the production directly would be more efficient. PR Is updated can you please recheck.
Thank you for the insight!